tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala

```diff
@@ -0,0 +1,156 @@
+    |
+    |The input file should have a single read group, with all fields from this read group copied to each new read group,
+    |besides the ID.  The read group ID for reads from a source molecule will be the original read group ID with the
+    |source molecule's ID appended (with a dash separator).  The sample name and library will have the molecular ID
+    |appended (also with a dash separator).
+    |
+    |An interval list can be given to filter output reads to a set of intervals.
+  """)
+class ReadGroupPerDuplexMolecularId
+(
+  @arg(flag='i', doc="Input SAM or BAM file (from GroupReadsByUmi).") val input: PathToBam,
+  @arg(flag='o', doc="Output SAM or BAM.") val output: PathToBam = Io.StdOut,
+  @arg(flag='r', doc="Reference sequence fasta file.") val ref: Option[PathToFasta] = None,
+  @arg(flag='l', doc="Optional list of intervals to reads to.") val intervals: Option[PathToIntervals] = None,
+  @arg(flag='T', doc="The output tag from UMI grouping.") val assignTag: String = ConsensusTags.MolecularId
+) extends FgStrTool with LazyLogging {
+
+  Io.assertReadable(input)
+  intervals.foreach(Io.assertReadable)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    // Pass 1: find all the molecular identifiers to use
+    logger.info("Extracting molecular ids")
+    val mids = {
+      val progress = ProgressLogger(this.logger, unit=5e5.toInt, verb="Extracted molecule ids")
+      val in       = SamSource(input, ref=ref)
+      val iterator = toIterator(in)
+      val ids      = new mutable.HashSet[Long]()
+      iterator.foreach { rec =>
+        ids += toMolecularId(rec, this.assignTag)
+        progress.record(rec)
+        if (progress.getCount % 5e5.toInt == 0) {
+          logger.info("# of molecules so far: " + ids.size)
+        }
+      }
+      in.safelyClose()
+      logger.info(s"Extracted ${ids.size} molecular ids across ${progress.getCount} raw reads")
+      ids.toSeq.sorted
+    }
+
+    // Pass 2: update the header and records
+    {
+      logger.info("Writing to output")
+      val in       = SamSource(input, ref=ref)
+      val iterator = toIterator(in)
+      val header   = toOutputHeader(in, mids)
+      val out      = SamWriter(output, header, ref=ref)
+      val rgId     = in.header.getReadGroups.get(0).getId
+      iterator.foreach { rec =>
+        rec("RG") = s"$rgId-${toMolecularId(rec, this.assignTag)}"
+        out += rec
+      }
+      in.safelyClose()
+      out.close()
+    }
+  }
+
+  /** Create the output header. */
+  private def toOutputHeader(in: SamSource, mids: Seq[Long]): SAMFileHeader = {
+    val header = in.header.clone()
+
+    val sourceReadGroup = header.getReadGroups.toList match {
+      case Nil       => throw new IllegalArgumentException(s"Found no read groups in ${this.input}")
+      case rg :: Nil => rg
+      case rgs       => throw new IllegalArgumentException(s"Found more than one read group in ${this.input}")
```

Comment:
I think this is fine for now, but you could presumably handle multiple where the library/sample are the same right?